working/macros/example/lib/checks_extensions.dart

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+// There is no public API exposed yet, the in-progress API lives here.
+import 'package:_fe_analyzer_shared/src/macros/api.dart';
+
+import 'util.dart';
+
+/// Generates extensions for the `checks` package for a list of types.
+///
+/// The extensions will be "on" the type `Subject<SomeType>`, for each given
+/// type.
+///
+/// Each extension will have a getter for each field in the type it is
+/// targetting, of the form `Subject<SomeFieldType>
+macro class ChecksExtensions implements LibraryTypesMacro {
+  final List<NamedTypeAnnotation> types;
+
+  const ChecksExtensions(this.types);
+
+  @override
+  Future<void> buildTypesForLibrary(
+      Library library, TypeBuilder builder) async {
+    // ignore: deprecated_member_use
+    final subject = await builder.resolveIdentifier(
+        Uri.parse('package:checks/checks.dart'), 'Subject');
+    final checksExtension = await builder.resolveIdentifier(
+        Uri.parse('package:macro_proposal/checks_extensions.dart'),
+        'ChecksExtension');
+    for (final type in types) {
+      if (type.typeArguments.isNotEmpty) {
+        throw StateError('Cannot generate checks extensions for types with '
+            'explicit generics');
+      }
+      final name = '${type.identifier.name}Checks';
+      builder.declareType(
+          name,
+          DeclarationCode.fromParts([
+            '@',
+            checksExtension,
+            '()',
+            'extension $name on ',
```

Comment:
This is interesting - it looks like the macro runs in two steps?

Does that mean that folks who want more customization can do something like

```dart
@ChecksExtension()
extension CustomName on Subject<SomeType> {}
```

Reply:
Yeah, it wouldn't have to work this way but that is imo the best way to write it. It would have to run in two phases regardless.

The macro could be smart enough to skip adding extensions that already exist though, so you could write custom ones and have it fill in the rest.

I am not sure all other macro use cases would want the 2nd macro (`@ChecksExtension()` here) to actually be visible so it is a little weird. It could actually exist in a `src` library, so that might be the solution "hiding" it. But it has to be public.